### PR TITLE
[JetBrains] Avoid shutting down terminals when closing the client

### PR DESCRIPTION
**Pull request by felladrin**
## Description

This PR changes the following:
- Avoid shutting down terminals when closing JetBrains Client.
- Open terminals for each Gitpod Task Terminal running and close them when JetBrains Client gest closed. We need to do this because clicking "Disconnect" is different from clicking "X" button from JetBrains Client Window [[1](https://github.com/gitpod-io/gitpod/issues/12126#issuecomment-1367325029)]. Now the terminals are behaving the same as in VS Code.
- Replace the use of `application.executeOnPooledThread` (thread) with `lifetime.launch` (coroutine).

## About the changes

Making the `GitpodTerminalService` class extend `Disposable` opens doors for us to use the `defineNestedLifetime()` method inside the class.
This method creates a LifetimeDefinition that is automatically terminated whenever the class is disposed.
`GitpodTerminalService` is created when a Controller Client connects and disposed when the client disconnects.
The LifetimeDefinition also allow us to launch coroutines (via `launch` method), and these coroutines are automatically terminated when the lifetime terminates.
This allows us to rely on the lifetime life-cycle to avoid running something in case the Controller Client disconnects.
For example: If we do the following, `<some_task>` won't be executed if the lifetime is terminated within 5 seconds after the coroutine is launched:
```
lifetime.launch {
    delay(5000)
    <some_task>
}
```
We currently have no way to know if a terminal widget was closed manually by the user or if it was closed due to the user closing JetBrains Client [[1](https://gitpod.slack.com/archives/C02BRJLGPGF/p1668774814801159)], so this allows us to prevent `<some_task>` [shutting down Gitpod Task Terminals, in our case] from running if JetBrains client has been closed. Gitpod Task Terminals are only shut down if the user clicks the "X" button in the terminal widget.

## Related Issue(s)

- Resolves https://github.com/gitpod-io/gitpod/issues/12126

## How to test

1. Open the preview environment of this PR and select Stable IntelliJ IDEA as your editor.
2. Start a new workspace from Gitpod's Repository, which has 5 terminals opened when it starts.
3. Close JetBrains Client by clicking "X" button.
4. Reconnect to the workspace and confirm if all the terminals remain open. Repeat it 3 times to ensure they never get closed unless you click "X" button on the terminal widget. (You can also use `gp tasks list` to know if Gitpod Terminals are in the "running" state)

## Release Notes

```release-note
Fixed an issue which caused Gitpod Terminals to be terminated when closing JetBrains Client.
```

## Werft options:

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [x] /werft with-large-vm
- [ ] /werft with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`

### CODE REVIEW

**Comment by felladrin** [comment]
Almost there. There are two cases to consider [[1](https://github.com/gitpod-io/gitpod/issues/12126#issuecomment-1367325029)]:
1. User can disconnect using the connection window from Gateway.
2. User can close the JetBrains Client, without stopping the workspace.

In the first case, the terminals remain open when the user gets back. In the second, terminals are all closed.

**Comment by felladrin** [comment]
Updated. Now it closes the terminal widgets (without shutting down the Gitpod Task Terminals) and reopens them when the client connects again.

File: components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/GitpodTerminalService.kt [hunk 1]
```diff
@@ -270,12 +276,23 @@ class GitpodTerminalService(project: Project) {
...
         })
     }
+
+    private fun closeTerminalWidgetWhenClientGetsClosed(
+            shellTerminalWidget: ShellTerminalWidget
+    ) {
+        lifetime.onTerminationOrNow {
+            shellTerminalWidget.close()
+        }
+    }
```
Comment (Siddhant-K-code) [inline components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/GitpodTerminalService.kt:297]: **Suggestion (non-blocking):** Making it more robust - 

```suggestion
    private fun closeTerminalWidgetWhenClientGetsClosed(
            shellTerminalWidget: ShellTerminalWidget
    ) {
        lifetime.onTerminationOrNow {
            try {
                shellTerminalWidget.close()
            } catch (e: Exception) {
                logger.error(e) { "Failed to close terminal widget" }
            }
        }
    }
```

File: components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/GitpodTerminalService.kt [hunk 2]
```diff
@@ -270,12 +276,23 @@ class GitpodTerminalService(project: Project) {
...
-                        TerminalOuterClass.ShutdownTerminalRequest.newBuilder()
-                                .setAlias(supervisorTerminal.alias)
-                                .build()
-                )
+                lifetime.launch {
+                    delay(5000)
+                    terminalServiceFutureStub.shutdown(
+                            TerminalOuterClass.ShutdownTerminalRequest.newBuilder()
+                                    .setAlias(supervisorTerminal.alias)
+                                    .build()
```
Comment (Siddhant-K-code) [inline components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/GitpodTerminalService.kt:284]: **Suggestion(non-blocking)**

```suggestion
                    try {
                        terminalServiceFutureStub.shutdown(
                                TerminalOuterClass.ShutdownTerminalRequest.newBuilder()
                                        .setAlias(supervisorTerminal.alias)
                                        .build()
                        )
                    } catch (e: StatusRuntimeException) {
                        LOG.info("Could not shutdown terminal ${supervisorTerminal.alias} due to ${e.message}")
                    }
```

Comment (felladrin) [reply components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/GitpodTerminalService.kt:284]: Good idea! Let me check if I can catch it like that.

**Review by andreafalzetti — APPROVED** [review]
Reviewed and tested, works as expected ✅ 

/hold

Unhold when you're ready

File: components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/GitpodTerminalService.kt (hunk 2 above)
Comment (felladrin) [reply components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/GitpodTerminalService.kt]: ✅  [Try-catch added](https://github.com/gitpod-io/gitpod/compare/bdd2bb67b20e37c1f8d7edb790512a2daaba5f47..bfdd9c3beeac644999acebff8d36a1d4d8a9c593).

**Pull request merged**